### PR TITLE
Fix unmarshaling of nested DS Dictionary

### PR DESCRIPTION
**Pull request by aparajit-pratap**
### Purpose

Fix issue with passing nested DesignScript.Builtin Dictionary to ZT node accepting generic Dictionary as argument.
JIRA link: https://jira.autodesk.com/browse/QNTM-4917

image

image

Zero touch node definitions used in this example:
```
 public static Dictionary<string, int> ReturnDictionary()
        {
            var dictionary = new Dictionary<string, int>();
            dictionary.Add("A", 1);
            dictionary.Add("B", 2);
            dictionary.Add("C", 3);
            dictionary.Add("D", 4);
            return dictionary;
        }
 [MultiReturn(new string[] {"H", "G", "F", "E"})]
        public static Dictionary<string, object> ReturnNestedDictionary()
        {
            var dict = ReturnDictionary();
            var dictionary = new Dictionary<string, object>();
            dictionary.Add("E", 1);
            dictionary.Add("F", 2);
            dictionary.Add("G", dict);
            dictionary.Add("H", false);
            return dictionary;
        }
 public static Dictionary<string, object> ReturnArbitraryDictionary()
        {
            var dict = ReturnDictionary();
            var dictionary = new Dictionary<string, object>();
            dictionary.Add("E", 1);
            dictionary.Add("F", new Dummy());
            dictionary.Add("G", dict);
            dictionary.Add("H", false);
            dictionary.Add("I", new object[] {1, 2, 3, 4, new DummyCollection()});

            return dictionary;
        }

        public static IDictionary ReturnIDictionary()
        {
            return ReturnDictionary();
        }

        public static IDictionary ReturnNestedIDictionary()
        {
            return ReturnNestedDictionary();
        }
public static IDictionary AcceptNestedDictionary(Dictionary<string, object> dictionary)
        {
            return dictionary;
        }
public static IDictionary AcceptIDictionary(IDictionary dictionary)
        {
            return dictionary;
        }
```
The following table shows the different combinations of Dictionary output (rows) being passed into node accepting Dictionary inputs (columns):

  | Dictionary<Key, Value> | IDictionary
-- | -- | --
Dictionary<Key, Value> | Y | Y
IDictionary | Y | Y
Nested Dictionary | Y | Y
Nested IDictionary | Y | Y
Dictionary w/ user-defined   type | Y | Y
IDictionary w/ user-defined   type | Y | Y
Arbitrary Dynamo Dictionary | Y | Y
Dictionary w/ List | Y | Y
Idictionary w/ List | Y | Y

### Declarations

Check these if you believe they are true

- [X] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI. 
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@saintentropy 

### FYIs

@jnealb

### CODE REVIEW

File: src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs [hunk 1]
```diff
@@ -533,7 +533,7 @@ private object ToIDictionary(StackValue dsObject, ProtoCore.Runtime.Context cont
             {
                 try
                 {
-                    targetDict[key] = Convert.ChangeType(d.ValueAtKey(key), valueType);
+                    targetDict[key] = d.ValueAtKey(key);
```
Comment (aparajit-pratap) [inline src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs]: I don't see any necessity for the type conversion in this case. This conversion was throwing the error for the input object (DesignScript.Builtin.Dictionary) not being an IConvertible in the first place.

**Comment by erfajo** [comment]
You can do it like this

    public static class Dictionary
    {
        static public IDictionary ReadDictionary(IDictionary dictionary)
        {
            return dictionary;
        }
    }

capture

**Comment by aparajit-pratap** [comment]
Thanks @erfajo. Yes, that's one way to do it but we are also trying to fix the generic dictionary case.

**Comment by alfarok** [comment]
@aparajit-pratap Changes look good to me.  Just to confirm the previous behavior still functioned as expected but the node accepting the generic dictionary as an input displayed in a warning state?

**Comment by aparajit-pratap** [comment]
Thanks @alfarok for the review and testing; I've added some more tests. Merging.

**Pull request merged**